### PR TITLE
SSO: Implement users table user invite revoking [Original]

### PR DESCRIPTION
**Pull request by agrullon95**
Fixes #

## Proposed changes:

* These changes build on the users table improvements from: https://github.com/Automattic/jetpack/pull/35158.
* The logic to revoke user invites has been added to the users table row actions for each user that has a 'pending invite'.

CleanShot 2024-01-25 at 22 49 56@2x

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

1. On an Atomic, JN, or whatever non-simple site, install Jetpack.
2. Make sure SSO is on, it should be on by default.
3. Sync this PR to your site.
4. Go to the users table (`wp-admin/users/php`).
5. It should have the status of the user's connection "Connected", "Invite", or "Pending invite".
6. Clicking "Invite" should invite the user and change their status to "Pending invite".
7. If you haven't already, add and invite a few new users to your site.
8. Hover the row actions and verify that rows with 'pending invite' only display the `Revoke invite` action.
9. Verify the 'Revoke invite' row action successfully revokes a user's site invitation.

### CODE REVIEW

File: projects/plugins/jetpack/modules/sso.php [hunk 1]
```diff
@@ -86,7 +86,140 @@ private function __construct() {
...
+				$url = add_query_arg(
+					array(
+						'jetpack-sso-revoke-user-invite' => 'failed',
+						'jetpack-sso-revoke-user-invite-error' => $reason,
+						'_wpnonce'                       => $nonce,
+					),
+					$ref
+				);
+				return wp_safe_redirect( $url );
+			}
```
Comment (alshakero) [inline projects/plugins/jetpack/modules/sso.php]: This should be moved up before checking if the invite exists.

File: projects/plugins/jetpack/modules/sso.php [hunk 2]
```diff
@@ -86,7 +86,140 @@ private function __construct() {
...
+				array(
+					'jetpack-sso-revoke-user-invite' => $body->deleted ? 'success' : 'failed',
+					'_wpnonce'                       => $nonce,
+				),
+				$ref
+			);
+			if ( ! $body->deleted ) {
+				$url = add_query_arg(
+					array(
+						'jetpack-sso-revoke-user-invite-error' => 'invalid-invite-revocation',
```
Comment (alshakero) [inline projects/plugins/jetpack/modules/sso.php:205]: Missing nonce.

File: projects/plugins/jetpack/modules/sso.php [hunk 3]
```diff
@@ -86,7 +86,140 @@ private function __construct() {
...
+	}
+
+	/**
+	 * Handles logic to revoke user invite.
+	 */
+	public function revoke_user_invite_to_wpcom() {
+		check_admin_referer( 'jetpack-sso-revoke-user-invite', 'revoke_invite_nonce' );
+		$nonce = wp_create_nonce( 'jetpack-sso-revoke-user-invite' );
+
+		if ( ! current_user_can( 'create_users' ) ) {
```
Comment (alshakero) [inline projects/plugins/jetpack/modules/sso.php]: The permission should be `promote_users` I think.

File: projects/plugins/jetpack/modules/sso.php [hunk 4]
```diff
@@ -86,7 +86,140 @@ private function __construct() {
...
+	 * Adds 'Revoke invite' link to user table row actions.
+	 *
+	 * @param array   $actions - User row actions.
+	 * @param WP_User $user_object - User object.
+	 */
+	public function jetpack_user_table_row_actions( $actions, $user_object ) {
+		$user_id            = $user_object->ID;
+		$has_pending_invite = self::has_pending_wpcom_invite( $user_id );
+
+		if ( current_user_can( 'create_users' ) && $has_pending_invite ) {
```
Comment (alshakero) [inline projects/plugins/jetpack/modules/sso.php:105]: Should probably be `promote_users`. Creating an invitation only requires that anyways.

**Review by alshakero — CHANGES_REQUESTED** [review]
Love it. Really clean.

File: projects/plugins/jetpack/modules/sso.php [hunk 5]
```diff
@@ -116,6 +249,34 @@ public function handle_invitation_results() {
...
+				case 'invalid-user':
+					return wp_admin_notice( __( 'Tried to revoke an invite for a user that doesn&#8217;t exist.', 'jetpack' ), array( 'type' => 'error' ) );
+				case 'invalid-email':
+					return wp_admin_notice( __( 'Tried to revoke an invite for a user that doesn&#8217;t have an email address.', 'jetpack' ), array( 'type' => 'error' ) );
+				case 'invalid-invite-revocation':
+					return wp_admin_notice( __( 'Tried to revoke an invite that doesn&#8217;t exist.', 'jetpack' ), array( 'type' => 'error' ) );
+				case 'invalid-permissions':
+					return wp_admin_notice( __( 'You don&#8217;t have permission to revoke invites.', 'jetpack' ), array( 'type' => 'error' ) );
+				default:
+					return;
```
Comment (alshakero) [inline projects/plugins/jetpack/modules/sso.php]: Can go.

File: projects/plugins/jetpack/modules/sso.php (hunk 4 above)
Comment (agrullon95) [reply projects/plugins/jetpack/modules/sso.php:105]: Updated to `promote_users`. Aligns better with usage in endpoint: fbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Szh%2Qcyhtvaf%2Svaivgr%2Svaivgr%2Qhfref.cuc%3Se%3Q94217699%2325-og

File: projects/plugins/jetpack/modules/sso.php (hunk 5 above)
Comment (agrullon95) [reply projects/plugins/jetpack/modules/sso.php]: Removed this default case.

File: projects/plugins/jetpack/modules/sso.php (hunk 2 above)
Comment (agrullon95) [reply projects/plugins/jetpack/modules/sso.php:205]: Added missing nonce.

**Comment by crisbusquets** [comment]
I'll check this now, @agrullon95!

**Comment by escapemanuele** [comment]
Is it ok to close this @agrullon95 ?

**Comment by agrullon95** [comment]
> Is it ok to close this @agrullon95 ?

@escapemanuele - Yes, im closing this now.

**Pull request closed without merging**